src/OpenDict.h

```diff
@@ -0,0 +1,351 @@
+// See the file "COPYING" in the main distribution directory for copyright.
+//Clustered Hashing, a variation of Robinhood Hashing/Open Addressing Hashing.
+
+#pragma once
+#include <vector>
+#include "Hash.h"
+#include <climits>
+class Dictionary;
+class DictEntry;
+class IterCookie;
+
+// Default number of hash buckets in dictionary.  The dictionary will
+// increase the size of the hash table as needed.
+#define HASH_MASK 0xFFFFffff //only lower 32 bits.
+
+//micros here are used to build different targets with -Dxxx to compare for performance or for debug purposes.
+//when incremental resizing, remapping, it remaps DICT_REMAP_ENTRIES at step.
+#ifndef DICT_REMAP_ENTRIES
+#define DICT_REMAP_ENTRIES 16 //2 for debug. 16 is best for release. move how many at a time when remapping.
+#endif//DICT_REMAP_ENTRIES
+
+#ifndef DICT_LOAD_FACTOR_BITS //give option to define it at command line
+#define DICT_LOAD_FACTOR_BITS  2//LOAD_FACTOR = 1 - 1/2^LOAD_FACTOR_BITS .75 is the optimal load factor.
+#endif //LOAD_FACTOR_BITS
+
+//when log2_buckets > DICT_THRESHOLD_BITS, DICT_LOAD_FACTOR_BITS becomes effective.
+//basicly it means if dict size < 2^DICT_THRESHOLD_BITS+n, size up only if necessary (the last entry can't hold new key).
+#ifndef DICT_THRESHOLD_BITS
+#define DICT_THRESHOLD_BITS 3
+#endif
+
+// of insertions.
+typedef enum { ORDERED, UNORDERED } dict_order;
+
+// Type for function to be called when deleting elements.
+typedef void (*dict_delete_func)(void*);
+
+// A dict_delete_func that just calls delete.
+extern void generic_delete_func(void*);
+// The value of an iteration cookie is the bucket and offset within the
+// bucket at which to start looking for the next value to return.
+#define TOO_FAR_TO_REACH 0xFFFF
+#define MAX_KEY_SIZE 0XFFFF
+struct DictEntry{//24 bytes. perfectly alligned. always own the key. but not the value.
```

Comment:
This was forward-declared as a `class` earlier, and the compiler complains about it now being redefined as a `struct`. It doesn't look like the forward-declaration is used at all, so I would redefine this as a `class` and remove the earlier declaration.